File: backtracking/generateParentheses.py
# https://leetcode.com/problems/generate-parentheses/
class Solution:
    def generateParenthesis(self, n: int) -> List[str]:
        res = []
        self.dfs(n, n, "", res)
        return res

    def dfs(self, leftRemain, rightRemain, path, res):
        if leftRemain > rightRemain or leftRemain < 0 or rightRemain < 0:
            return  # backtracking
        if leftRemain == 0 and rightRemain == 0:
            res.append(path)
            return
        self.dfs(leftRemain-1, rightRemain, path+"(", res)
        self.dfs(leftRemain, rightRemain-1, path+")", res)

File: backtracking/generateParentheses.js
// https://leetcode.com/problems/generate-parentheses/
const go = (leftRemain, rightRemain, path, res) => {
  if (leftRemain > rightRemain) return; // The number of '(' should be always >= ')'

  if (leftRemain === 0 && rightRemain === 0) {
    res.push(path);
    return;
  }

  if (leftRemain > 0) go(leftRemain - 1, rightRemain, path + "(", res);
  if (rightRemain > 0) go(leftRemain, rightRemain - 1, path + ")", res);
};

/**
 * @param {number} n
 * @return {string[]}
 */
const generateParenthesis = (n) => {
  const res = [];
  go(n, n, "", res);
  return res;
};
